api/deployment/DeploymentRestHandler.go

```diff
@@ -0,0 +1,122 @@
+	"io/ioutil"
+	"net/http"
+	"time"
+)
+
+type DeploymentRestHandler interface {
+	CreateChartFromFile(w http.ResponseWriter, r *http.Request)
+}
+
+type DeploymentRestHandlerImpl struct {
+	Logger             *zap.SugaredLogger
+	userAuthService    user.UserService
+	enforcer           casbin.Enforcer
+	validator          *validator.Validate
+	refChartDir        pipeline.RefChartDir
+	chartService       pipeline.ChartService
+	chartRefRepository chartRepoRepository.ChartRefRepository
+}
+
+func NewDeploymentRestHandlerImpl(Logger *zap.SugaredLogger, userAuthService user.UserService, enforcer casbin.Enforcer, validator *validator.Validate,
+	refChartDir pipeline.RefChartDir, chartService pipeline.ChartService, chartRefRepository chartRepoRepository.ChartRefRepository) *DeploymentRestHandlerImpl {
+	return &DeploymentRestHandlerImpl{
+		Logger:             Logger,
+		userAuthService:    userAuthService,
+		enforcer:           enforcer,
+		validator:          validator,
+		refChartDir:        refChartDir,
+		chartService:       chartService,
+		chartRefRepository: chartRefRepository,
+	}
+}
+
+func (handler *DeploymentRestHandlerImpl) CreateChartFromFile(w http.ResponseWriter, r *http.Request) {
+	userId, err := handler.userAuthService.GetLoggedInUser(r)
+	if userId == 0 || err != nil {
+		common.WriteJsonResp(w, err, nil, http.StatusUnauthorized)
+		return
+	}
+
+	token := r.Header.Get("token")
+	if ok := handler.enforcer.Enforce(token, casbin.ResourceGlobal, casbin.ActionUpdate, "*"); !ok {
+		common.WriteJsonResp(w, errors.New("unauthorized"), nil, http.StatusForbidden)
+		return
+	}
+
+	file, fileHeader, err := r.FormFile("BinaryFile")
+	if err != nil {
+		handler.Logger.Errorw("request err, File parsing error", "err", err, "payload", file)
+		common.WriteJsonResp(w, err, nil, http.StatusBadRequest)
+		return
+	}
+
+	//chartName := strings.Split(fileHeader.Filename, ".")
+	if err := r.ParseForm(); err != nil {
+		handler.Logger.Errorw("request err, Corrupted form data", "err", err, "payload", file)
+		common.WriteJsonResp(w, err, nil, http.StatusBadRequest)
+		return
+	}
+
+	err = handler.chartService.ValidateFileUploaded(fileHeader.Filename)
+	if err != nil {
+		handler.Logger.Errorw("request err, Unsupported format", "err", err, "payload", file)
+		common.WriteJsonResp(w, errors.New("Unsupported format file is uploaded, please upload file with .tar.gz extension"), nil, http.StatusBadRequest)
+		return
+	}
+
+	fileBytes, err := ioutil.ReadAll(file)
+	if err != nil {
+		handler.Logger.Errorw("request err, File parsing error", "err", err, "payload", file)
+		common.WriteJsonResp(w, err, nil, http.StatusBadRequest)
+		return
+	}
+
+	chartLocation, chartName, chartVersion, err := handler.chartService.ExtractChartIfMissing(fileBytes, string(handler.refChartDir), "")
+
+	if err != nil {
+		common.WriteJsonResp(w, err, nil, http.StatusBadRequest)
+		return
+	}
+
+	exists, err := handler.chartRefRepository.DataExists(chartName, chartVersion)
```

Comment:
check this before extracting or atleast before copying otherwise this ll create issue, even though it is rejecting the upload it has already copied file in final destination